File: veytrix/frontend/src/services/logger.service.ts
import { supabase, supabaseAdmin } from '../lib/supabase';

export type LogLevel = 'info' | 'warning' | 'error';

export interface SystemLogEntry {
  id?: string;
  service: string;
  level: LogLevel;
  message: string;
  metadata?: Record<string, any>;
  created_at?: string;
}

class LoggerService {
  private async createLog(service: string, level: LogLevel, message: string, metadata?: Record<string, any>): Promise<void> {
    const payload = {
      service,
      level,
      message,
      metadata: metadata || {},
      created_at: new Date().toISOString(),
    };

    // Always output to developer console
    const logPrefix = `[${level.toUpperCase()}] [${service}]`;
    if (level === 'error') {
      console.error(logPrefix, message, metadata);
    } else if (level === 'warning') {
      console.warn(logPrefix, message, metadata);
    } else {
      console.log(logPrefix, message, metadata);
    }

    try {
      const client = supabaseAdmin || supabase;
      if (client) {
        const { error } = await client.from('system_logs').insert([payload]);
        if (error) {
          // If system_logs table has RLS or connection issue, fall back gracefully
          console.warn('System log persistence warning:', error.message);
        }
      }
    } catch (e) {
      console.warn('Could not persist system log to database:', e);
    }
  }

  public async logInfo(service: string, message: string, metadata?: Record<string, any>): Promise<void> {
    return this.createLog(service, 'info', message, metadata);
  }

  public async logWarning(service: string, message: string, metadata?: Record<string, any>): Promise<void> {
    return this.createLog(service, 'warning', message, metadata);
  }

  public async logError(service: string, message: string, metadata?: Record<string, any>): Promise<void> {
    return this.createLog(service, 'error', message, metadata);
  }
}

export const loggerService = new LoggerService();
export default loggerService;

File: veytrix/backend/app/services/logger_service.py
import logging
import asyncio
from typing import Optional, Dict, Any
from app.core.supabase import get_supabase_client

logger = logging.getLogger("veytrix.backend_logs")

async def log_backend_event(service: str, level: str, message: str, metadata: Optional[Dict[str, Any]] = None) -> None:
    """
    Logs backend execution events into public.backend_logs table for real-time monitoring in Command Center.
    """
    payload = {
        "service": service,
        "level": level.lower(),
        "message": message,
        "metadata": metadata or {}
    }
    
    # Standard logger output
    log_msg = f"[BACKEND] [{level.upper()}] [{service}] {message}"
    if level.lower() == "error":
        logger.error(log_msg)
    elif level.lower() == "warning":
        logger.warning(log_msg)
    else:
        logger.info(log_msg)

    # Persist to Supabase backend_logs database table
    try:
        supabase = get_supabase_client()
        if supabase:
            loop = asyncio.get_event_loop()
            await loop.run_in_executor(
                None,
                lambda: supabase.table("backend_logs").insert(payload).execute()
            )
    except Exception as e:
        logger.warning(f"Could not persist backend_logs row: {str(e)}")

def log_info(service: str, message: str, metadata: Optional[Dict[str, Any]] = None) -> None:
    try:
        asyncio.create_task(log_backend_event(service, "info", message, metadata))
    except Exception:
        pass

def log_warning(service: str, message: str, metadata: Optional[Dict[str, Any]] = None) -> None:
    try:
        asyncio.create_task(log_backend_event(service, "warning", message, metadata))
    except Exception:
        pass

def log_error(service: str, message: str, metadata: Optional[Dict[str, Any]] = None) -> None:
    try:
        asyncio.create_task(log_backend_event(service, "error", message, metadata))
    except Exception:
        pass
